### PR TITLE
validate drpolicy's s3 buckets and cluster roles exist on each reconcile attempt

### LINKED ISSUE #318: update drpolicy validated status condition's message if it changes
A drpolicy's s3 profile may be invalid for several reasons.  Its status condition's message indicates why an s3 profile's connection failed.  If the reason why an s3 profile connection is invalid changes, the status condition's message is not updated but it should be.

A drpolicy may have several s3 profiles.  Each s3 profile must be valid for a drpolicy's validated status condition to be true.  The validated status condition's message includes the name of the first profile found with an s3 connectivity failure.  If one profile is invalid and becomes valid, but another is found to be invalid the message is not updated but should be.

This issue was found in review of pull request #310 , but not addressed before it was merged:
https://github.com/RamenDR/ramen/pull/310/files#r731801571

### PR DESCRIPTION
**Pull request by hatfieldbrian**
- for each drpolicy reconciliation attempt,
  - validate:
    - each s3 store's bucket exists
    - drpolicy's finalizer exists
    - each cluster's cluster roles manifest work exists
  - update validated status condition if its status, reason, or message differs
    - closes #318

### CODE REVIEW

File: controllers/drpolicy_controller.go [hunk 1]
```diff
@@ -71,18 +71,24 @@ func (r *DRPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 	case true:
 		log.Info("create/update")
 
+		validatedSetTrue, validatedSetFalse := validatedSetFuncs(ctx, drpolicy, r.Client, log)
```
[the remark this answers is not in the document]
Comment (hatfieldbrian) [reply controllers/drpolicy_controller.go]: https://github.com/RamenDR/ramen/pull/328/commits/013d0beba82fc24e9f92947286377ac3c47b6abe refactors with set true and false routines calling a common routine

File: controllers/drpolicy_controller.go [hunk 2]
```diff
@@ -98,66 +104,138 @@ func (r *DRPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
-			return nil
+			return func(reason, message string) error {
+					log.Info("valid -> valid")
+
+					return nil
+				},
+				func(reason string, err error) error {
+					log.Info("valid -> invalid")
+
+					return error1stUnless2ndNotNil(
```
[the remark this answers is not in the document]
Comment (hatfieldbrian) [reply controllers/drpolicy_controller.go]: The first error should always be non-nil.  The intent is to call `statusConditionUpdate` whether the first error is nil or not.  If validation fails and then a corresponding status update fails, the latter error trumps the former.

File: controllers/drpolicy_controller.go [hunk 3]
```diff
@@ -98,66 +104,134 @@ func (r *DRPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
+}
+
+func (b *bar) statusConditionUpdateIfReasonOrMessageDiffers(
+	condition *metav1.Condition,
+	reason string,
+	err error,
+	logMessage string,
+) error {
+	message := err.Error()
+	if condition.Reason == reason && condition.Message == message {
```
Comment (ShyamsundarR) [inline controllers/drpolicy_controller.go:192]: If condition generation is older than resource generation, update status.

**Review by ShyamsundarR — COMMENTED** [review]
The various functions are a bit confusing but the logic is fine as it progresses :)

File: controllers/drpolicy_controller.go (hunk 3 above)
Comment (hatfieldbrian) [reply controllers/drpolicy_controller.go:192]: I presume that applies even if the status, message, or reason do not change?

Comment (ShyamsundarR) [reply controllers/drpolicy_controller.go:192]: That's true.

Comment (hatfieldbrian) [reply controllers/drpolicy_controller.go]: commit https://github.com/RamenDR/ramen/pull/328/commits/7d256f925b4c25ce150b8eb9bf1bd3f733a39075 updates status condition if object's generation changes

**Review by BenamarMk — COMMENTED** [review]
Just a general comment about anonymous functions.  Too many of them will soon make the code hard to read.  Right now, the DRPolicy code is small, but once it grows and if we keep adding these anonymous functions (as the paten suggests), the code will become hard to read and understand. I prefer named functions so that it reminds me why I am calling that function.  Whereas an anonymous function does not, and you have to always read and understand what it is doing no matter how many times you look at it.  Let alone you can't reuse it, you can't test it separately, and more.
As far as *foo/bar*, I think they don't have a place in a production code. You can name 'bar' as `SetValidator` for example, but `bar` does not tell me anything.

File: controllers/drpolicy_controller.go [hunk 4]
```diff
@@ -98,69 +103,140 @@ func (r *DRPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
+			return error1stUnless2ndNotNil(
+				err,
+				b.statusConditionAppend(ramen.DRPolicyValidated, metav1.ConditionFalse, reason, message),
+			)
+		},
+	)
+}
 
-			return conditionAppend(metav1.ConditionTrue, reason, message)
+func (b *bar) validatedSet(foo1, foo2 func(*metav1.Condition) error, foo3 func() error) error {
```
[the remark this answers is not in the document]
Comment (hatfieldbrian) [reply controllers/drpolicy_controller.go]: commit https://github.com/RamenDR/ramen/pull/328/commits/e66f7dc9b23c9a918c7403e1ed015ad7320bdd39 refactors the status conditions updates

**Review by BenamarMk — APPROVED** [review]
LGTM. Thank you @hatfieldbrian for the update.

**Review by ShyamsundarR — APPROVED** [review]

**Pull request merged**